### ISSUE TITLE
# [Epic #115] AI Chat Integration (Teach Mode & Act Mode Dual-Channel Architecture)

### ISSUE THREAD

**Issue by LukasLiss** [LukasLiss/totem-tool#115]
## Goal
Embed an interactive AI chat assistant into the `totem-tool` UI, enabling process mining users to ask natural language questions about active event logs, receive step-by-step guided visual tours of analysis features (**Teach Mode**), and execute automated dashboard manipulations or navigation actions (**Act Mode**) with low perceived latency via token streaming.
## Out of Scope
- Training or fine-tuning AI models on private user event logs.
- Cross-session chat history persistence in the database (chat state persists only within active browser session storage).
---
## 🏗️ Architectural Overview & Mental Model

The AI Chat Assistant relies on a **Dual-Channel Communication Architecture**:

```mermaid
flowchart TD
    UB[User Browser]
    AST[backend/assistant/<br/>Agentic Brain & Teach Mode RAG]
    AGT[backend/agent/<br/>Real-Time WS Bus & Session Registry]
    MCP[backend/mcp_server/<br/>18 Tools & Policy Rules]

    UB -->|HTTP POST Stream/Chat| AST
    UB <-->|WebSocket Live Wire Commands| AGT
    AST -->|Tool Calls| MCP
    MCP -.->|Commands & Results| AGT
```

### Channel Division Matrix

| Channel | Entry Points (`Backend` ↔ `Frontend`) | Responsibilities / Carried Data |
| :--- | :--- | :--- |
| **HTTP Chat** | `assistant/views.py` ↔ `api/assistantApi.ts` | User prompt payload, LLM text streaming, Teach Mode tour highlight paths, RAG knowledge context, and pending confirmation action chips. |
| **WebSocket** | `agent/consumers.py` ↔ `agent/AgentBridge.tsx` | Real-time bidirectional execution wire driving frontend UI actions (`navigate`, `set_view_mode`, `click`, `create_dashboard`, `get_ui_state`). |

---

#### 1. Agentic Brain & Teach Mode (`backend/assistant/`)
* **`views.py` (`ChatView.post`)**: Main agentic orchestration loop, context ingestion, tool routing, and streaming HTTP responses.
* **`llm.py`**: Anthropic SDK integration wrapper (`complete()`), handling text streaming and `tool_use` / `tool_result` block parsing.
* **`prompts.py` (`build_system_prompt()`)**: System prompt construction including base persona instructions, current UI state context, and the `data-tour-id` element catalogue.
* **`tools.py`**: Tool definitions for Teach Mode guidance (e.g., `show_path` schema and payload sanitization).
* **`urls.py`**: Route mounting for `/api/assistant/chat/`.
* **`knowledge/help.md`**: Embedded system documentation and user guide injected into prompts.
* **`knowledge/retriever.py`**: BM25 document retrieval system selecting top-3 relevant knowledge chunks per prompt.

#### 2. Act Mode Tool Server & Policy (`backend/mcp_server/`)
* **`server.py`**: Collection of ~18 executable tools (`TOOLS`, `TOOL_SPECS`, `call_tool()`), JSON schema definitions, and ORM interfaces.
* **`policy.py`**: Security policy enforcement categorizing tools into non-destructive read actions vs. `MUTATING` / `REQUIRES_FRONTEND` actions requiring explicitly confirmed user chips.

#### 3. WebSocket Command Bus (`backend/agent/`)
* **`consumers.py` (`AgentConsumer`)**: Django Channels WebSocket consumer handling live command/result protocols (`run_command()`).
* **`registry.py`**: Global session manager mapping authenticated `user_id → WebSocket connection`.
* **`jwt_ws_middleware.py`**: Custom JWT authentication middleware validating WS queries (`?token=`).

#### 4. Services & Configuration (`backend/api/services/` & `totem_backend/`)
* **`api/services/dashboard.py`**: Dashboard ORM service layer (creation, component insertion, auto-layout packing, ownership checks).
* **`totem_backend/settings.py`**: Configuration flags (`ANTHROPIC_API_KEY`, `ASSISTANT_MODEL`, `ASSISTANT_AGENT_ENABLED`).
* **`totem_backend/asgi.py`**: Protocol type router routing HTTP and WebSocket requests.

---

#### 1. Chat Panel UI (`frontend/src/components/chat/`)
* **`ChatWidget.tsx`**: Floating collapsible drawer UI widget, mode toggles (Teach Mode vs. Act Mode), and `localStorage` settings management.
* **`MessageList.tsx` & `MessageInput.tsx`**: Message bubble stream renderer (Markdown supported) and auto-resizing text input.
* **`PendingActions.tsx`**: Interactive confirmation chips for user approval of mutating backend tools.
* **`useAssistantContext.ts`**: Hook bundling active client context (route, selected event log ID, view mode, sidebar state) into chat payloads.

#### 2. Client API & WebSocket Agent Bridge (`frontend/src/api/` & `frontend/src/agent/`)
* **`api/assistantApi.ts`**: HTTP client wrapper (`sendChat()`) for request posting and response stream reading.
* **`agent/AgentBridge.tsx`**: WebSocket connection lifecycle manager, connection indicator state context, and dependency injection bridge.
* **`agent/handlers.ts`**: Client-side command handler registry mapping received agent commands (`navigate`, `set_view_mode`, `click`, `get_ui_state`) to React state updates.
* **`agent/types.ts`**: TypeScript definitions for wire frames (`Command`, `Result`, `Notice`).

#### 3. Teach Mode Guided Tour (`frontend/src/tour/`)
* **`TourController.tsx`**: State machine managing step-by-step UI tour flows and user interaction delegation.
* **`tourIds.ts`**: Centralized catalogue of `data-tour-id` attributes mirroring backend valid tour target IDs.
* **`Highlighter.tsx`**: Overlay renderer placing visual spotlights and hint callouts on target UI elements.

#### 4. UI Contexts & App Manipulation
* **`contexts/SelectedFileContext.tsx`**: Context providing active event log details to the assistant.
* **`contexts/DashboardContext.tsx`**: View mode state manager (`overview` | `analysis` | `dashboard`).
* **`contexts/SidebarGroupsContext.tsx`**: Control state for sidebar navigation groups.
* **`components/grid.tsx` & `app-sidebar.tsx`**: Listeners for `totem:refresh-dashboard` custom events to auto-update UI components.

---

### Sub-Task 1: `#116` - Spike: Architecture, API Contracts & Pre-flight Coordination
- [ ] Define HTTP request/response schemas for `/api/assistant/chat/`.
- [ ] Define WebSocket JSON frame specification for `agent/` live wire.
- [ ] Finalize standard JSON Schemas for MCP server tools (~18 tools).

### Sub-Task 2: `#117` - Frontend: Wire AI Chat Panel into UI with Streaming Responses
- [ ] Create collapsible drawer component `ChatWidget.tsx` in `frontend/src/components/chat/`.
- [ ] Implement SSE / Fetch ReadableStream reader for token streaming in `assistantApi.ts`.
- [ ] Add Markdown response rendering (with code highlight and list formatting).
- [ ] Implement `PendingActions.tsx` for mutation action approval chips.

### Sub-Task 3: `#118` - Backend: Assistant Agentic Brain & RAG Retriever
- [ ] Implement `ChatView.post` agentic loop in `backend/assistant/views.py`.
- [ ] Create Anthropic API wrapper `llm.py` with streaming support.
- [ ] Construct system prompt builder (`prompts.py`) with `data-tour-id` catalogue.
- [ ] Build BM25 doc retriever `retriever.py` over `backend/assistant/knowledge/help.md`.

### Sub-Task 4: `#119` - Backend: MCP Tool Server & Confirmation Policy Framework
- [ ] Define ~18 process mining & dashboard management tools in `backend/mcp_server/server.py`.
- [ ] Implement security policies in `backend/mcp_server/policy.py` (`MUTATING`, `REQUIRES_FRONTEND`).
- [ ] Connect dashboard service ORM operations in `api/services/dashboard.py`.

### Sub-Task 5: `#120` - Real-time WebSocket Command Bus & Agent Bridge
- [ ] Implement `AgentConsumer` and session registry in `backend/agent/`.
- [ ] Build JWT WebSocket authentication middleware (`jwt_ws_middleware.py`).
- [ ] Implement client-side `AgentBridge.tsx` and command handler registry (`handlers.ts`).

### Sub-Task 6: `#121` - Teach Mode Guided Tour & UI Highlighting System
- [ ] Build guided tour state machine `TourController.tsx` in `frontend/src/tour/`.
- [ ] Implement DOM element spotlight callouts with `Highlighter.tsx`.
- [ ] Sync `tourIds.ts` with backend `prompts.py` validation set.

### Sub-Task 7: `#122` - End-to-End Integration, Context Ingestion & Verification
- [ ] Wire active file & dashboard state via `useAssistantContext.ts`.
- [ ] Verify automatic UI refreshes via `totem:refresh-dashboard` custom events.
- [ ] Run backend unit tests and frontend E2E validation.

---

## ✅ Acceptance Criteria & Definition of Done

1. **Streaming UX**: Token-by-token text streaming achieves < 800ms Time-To-First-Token (TTFT).
2. **Teach Mode**: Asking "How do I filter variants?" highlights the correct sidebar item and triggers a visual callout tooltip.
3. **Act Mode**: Prompts like "Create a dashboard for process bottlenecks" invoke backend MCP tools, stream progress, and automatically update the frontend grid via WebSocket commands.
4. **Safety Policy**: Any mutating operation (e.g. deletion, overwriting layout) prompts an interactive confirmation chip before execution.
5. **Session Isolation**: Chat context is scoped strictly to the current active session and user account.
6. **Code Quality**: All backend test suites pass (`pytest`), and TypeScript types compile cleanly without errors.
